platform/CriticalSectionLock.h

```diff
@@ -34,35 +35,68 @@ namespace mbed {
   *     }
   *     // interrupts will be restored to their previous state
   * }
+  * void f() {
+  *     CriticalSectionLock cs(false);
+  *     // some code here
+  *
+  *     cs.lock();
+  *        // Code in this block will run with interrupts disabled
+  *     cs.unlock();
+  *     // interrupts will be restored to their previous state
+  *
+  *     cs.lock();
+  *        // Code in this block will run with interrupts disabled
+  *     cs.unlock();
+  *     // interrupts will be restored to their previous state
+  * }
   * @endcode
   */
 class CriticalSectionLock {
 public:
-    CriticalSectionLock() 
+    /** Create and initialize a CriticalSectionLock object.
+     * @param lock  critical section lock state after initialization
+     */
+    CriticalSectionLock(bool lock = true) : _locked(false)
     {
-        core_util_critical_section_enter();
+        if (lock) {
+            core_util_critical_section_enter();
+            _locked = true;
+        }
     }
 
     ~CriticalSectionLock() 
     {
-        core_util_critical_section_exit();
+        if (_locked) {
+            core_util_critical_section_exit();
+        }
     }
 
     /** Mark the start of a critical section
      *     
      */
     void lock()
     {
-        core_util_critical_section_enter();
+        MBED_ASSERT(false == _locked);
```

Comment:
Why is recursive locking not supported

Reply:
The idea was to prevent user from `lock/unlock` mismatch e.g. calling more locks then unlocks and  vice versa. The recursion could be handled by one CriticalSectionLock instance on each level

Reply:
Mismatch is taken care of in core_util_critical_section_enter / core_util_critical_section_exit functions.
Counter `interrupt_enable_counter` is maintained to keep track